examples/c++/userPropagator.cpp

```diff
@@ -0,0 +1,420 @@
+
+    model(unsigned cnt) : cnt(cnt), values(nullptr) {
+        values = (unsigned*) malloc(sizeof(unsigned) * cnt);
+    }
+
+    ~model() {
+        free(values);
+        values = nullptr;
+    }
+
+    unsigned &operator[](const int id) {
+        return values[id];
+    }
+
+    void set(unsigned pos, unsigned value) const {
+        values[pos] = value;
+    }
+
+    bool operator==(const model &other) const {
+        if (cnt != other.cnt) {
+            return false;
+        }
+        for (unsigned i = 0; i < cnt; i++) {
+            if (values[i] != other.values[i]) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    bool operator!=(const model &other) const {
+        return !this->operator==(other);
+    }
+
+    model* clone() const {
+        model *newModel = new model(cnt);
+        memcpy(newModel->values, values, sizeof(unsigned) * cnt);
+        return newModel;
+    }
+};
+
+struct model_hash_function {
+    std::size_t operator()(const model &m) const {
+        size_t hash = 0;
+        for (unsigned i = 0; i < m.cnt; i++) {
+            hash *= m.cnt;
+            hash += m.values[i];
+        }
+        return hash;
+    }
+};
+
+class user_propagator : public z3::user_propagator_base {
+
+protected:
+
+    std::unordered_map<unsigned, unsigned> *id_mapping;
+    model *currentModel;
+    std::unordered_set<model, model_hash_function> modelSet;
+    std::vector<unsigned> fixedValues;
+    std::stack<unsigned> fixedCnt;
+    unsigned board;
+
+    int solutionId = 1;
+
+public:
+
+    int getModelCount() const {
+        return solutionId - 1;
+    }
+
+    void final() final {
+        this->conflict((unsigned) fixedValues.size(), fixedValues.data());
+        if (modelSet.find(*currentModel) != modelSet.end()) {
+            WriteLine("Got already computed model");
+            return;
+        }
+        Write("Model #" << solutionId << ":\n");
+        solutionId++;
+        for (int i = 0; i < fixedValues.size(); i++) {
+            unsigned id = fixedValues[i];
+            WriteLine("q" + to_string((*id_mapping)[id]) + " = " + to_string((*currentModel)[id]));
+        }
+        modelSet.insert(*currentModel);
+        WriteEmptyLine;
+
+        currentModel = currentModel->clone();
+    }
+
+    static unsigned bvToInt(z3::expr e) {
+        return (unsigned)e.get_numeral_int();
+    }
+
+    void fixed(unsigned id, z3::expr const &e) override {
+        fixedValues.push_back(id);
+        unsigned value = bvToInt(e);
+        currentModel->set((*id_mapping)[id], value);
+    }
+
+    user_propagator(z3::solver *s, std::unordered_map<unsigned, unsigned> *idMapping, unsigned board)
+            : user_propagator_base(s), id_mapping(idMapping), board(board) {
+
+        currentModel = new model(board);
+
+        this->register_fixed();
+        this->register_final();
+    }
+
+    ~user_propagator() {
+        delete currentModel;
+        currentModel = nullptr;
```

Comment:
setting to nullptr is redundant. C++ has scoped ptr that does the free automatically.